subsys/storage/flash_map: DT independent flash area access API

de-nordic opened the pull request:
subsys/storage/flash_map: DT independent flash area access API

This commit provides DT independent flash area access API. This API
is supposed to provide compile time generated flash area access and
information, the same way the DT API did, but that is easier to replace
and using labels instead of identifiers

Enums are generated as identifiers.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

Note: This changes are placed on top of changes to gen_defines and util.h that are drafted here:
https://github.com/zephyrproject-rtos/zephyr/pull/24514
https://github.com/zephyrproject-rtos/zephyr/pull/24847
https://github.com/zephyrproject-rtos/zephyr/pull/24848
~~https://github.com/zephyrproject-rtos/zephyr/pull/24503~~

de-nordic committed:
subsys/storage/flash_map: DT independent flash area access API

This commit provides DT independent flash area access API. This API
is supposed to provide compile time generated flash area access and
information, the same way the DT API did, but that is easier to replace
and using labels instead of identifiers.

Enums are generated as identifiers.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
devicetree: gen_defines: Generate list of nodes' children

For each node, a list of identifiers of children nodes is generated
in form of:
DT_S_<path_to_node>_CHILD_IDENTIFIERS <id0>[, <id1>[, ...]]

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

mbolivar-nordic commented on subsys/storage/flash_map/flash_map_default.c:
```diff
@@ -7,15 +7,17 @@
 
 const struct flash_area default_flash_map[] = {
-	UTIL_LISTIFY(DT_FLASH_AREA_NUM, FLASH_AREA_FOO, ~)
+	UTIL_FOR_LIST(FLASH_AREA_GEN, FLASH_AREA_LIST)
```
I checked the preprocessor output with MACRO_MAP instead and it's the same, so I think UTIL_FOR_LIST is not needed

de-nordic replied:
Sorry, somehow I missed the `MACRO_MAP` thing. Btw. is counterintuitive that it takes all arguments under `...` instead of having fixed argument for function, followed by `...`.

de-nordic replied:
Moved to `MACRO_MAP` macros.

mbolivar-nordic commented on include/flash_pm.h:
```diff
@@ -0,0 +1,32 @@
+/* Number of flash areas */
+#define FLASH_AREA_N UTIL_LIST_LEN(FLASH_AREA_LIST)
+
+#define FLASH_AREA_ID(lbl) __dt_flash_defined_partition__##lbl
```
```suggestion
#define FLASH_AREA_ID(lbl) flash_area_id_##lbl
```

Double underscores are reserved for the implementation in C++, so we're trying to avoid using them.

de-nordic replied:
ok

mbolivar-nordic commented on include/flash_pm.h:
```diff
@@ -0,0 +1,32 @@
+
+/* Generate sequential identifiers for flash areas. */
+#define _FLASH_AREA_ID_GEN(lbl) FLASH_AREA_ID(lbl),
+enum __dt_flash_defined_partitions {
```
```suggestion
enum flash_area_id {
```

Same here

de-nordic replied:
ok

mbolivar-nordic commented on include/flash_pm.h:
```diff
@@ -0,0 +1,32 @@
+
+#define FLASH_AREA_OFFSET(lbl)	DT_REG_ADDR(DT_NODELABEL(lbl))
+#define FLASH_AREA_SIZE(lbl)	DT_REG_SIZE(DT_NODELABEL(lbl))
+#define FLASH_AREA_PTABLE(lbl)	DT_PARENT(DT_NODELABEL(lbl))
```
Can you explain the need for these macros to be in the API header instead of the default implementation file?

de-nordic replied:
They are supposed to be provided for a programmer to access FLASH_AREA information.

mbolivar-nordic reviewed (commented):
Thanks for this. That's roughly along the lines I was wondering if we could go also.

de-nordic commented on include/flash_pm.h:
```diff
@@ -0,0 +1,32 @@
+#define DEFAULT_DT_FIXED_PARTITIONS DT_INST(0, fixed_partitions)
+
+/* Right now only DT definitions are supported, and required */
+#if DT_NODE_HAS_COMPAT(DEFAULT_DT_FIXED_PARTITIONS, fixed_partitions)
```
Removed.

pabigot commented on include/flash_pm.h:
```diff
@@ -0,0 +1,46 @@
+#define ZEPHYR_INCLUDE_FLASH_PM_H_
+
+
+#define FLASH_AREA DT_CHOSEN(zephyr_flash)
```
Based on prefix these are related to the material that's defined in `<storage/flash_map.h>`.  Is there a reason they should not move there?

In any case the name `flash_pm.h` suggests it's related to power management for flash, which it isn't.

de-nordic replied:
I will move them.

mbolivar-nordic replied:
I spent some time trying to do this today and I believe there's a blocker here: this seems to lose the ability to work with multiple flash devices in the system, each with their own partitions.

Does the code need to be iterating over every node with compat `fixed_partitions` instead? Basically something like this:

```
/* in devicetree.h */
#define DT_FOREACH_NODE(compat, inst_expr) \
	COND_CODE_1(DT_HAS_COMPAT(compat), \
		    (UTIL_CAT(DT_FOREACH_INST_, compat)(inst_expr)), ())
// ... then redefine DT_INST_FOREACH(inst_expr) to be DT_FOREACH_NODE(DT_DRV_COMPAT, inst_expr)

/* flash_map.h */
#define GEN_AREAS_FOR_PARTITIONS_NODE(partition_inst) ...
enum flash_areas {
        DT_FOREACH_NODE(fixed_partitions, GEN_AREAS_FOR_PARTITIONS_NODE)
};
```

Where `GEN_AREAS_FOR_PARTITIONS_NODE(partition_inst)` works basically how `MACRO_MAP(FLASH_AREA_ID_GEN, FLASH_AREA_PARTITIONS_LIST)` is being used here, except it works on `DT_CHILDREN(DT_INST(partition_inst, fixed_partitions))` instead.

pabigot reviewed (changes requested):
Please move the macro definitions into `<storage/flash_map.h>` unless there's a good reason not to.

I remember there was some discussion about two APIs serving the same general purpose, with one of them preferred over the other.  If `<storage/flash_map.h>` is not the preferred one then more thought is required.

hakonfam commented:
> Please move the macro definitions into `<storage/flash_map.h>` unless there's a good reason not to.
> 
> I remember there was some discussion about two APIs serving the same general purpose, with one of them preferred over the other. If `<storage/flash_map.h>` is not the preferred one then more thought is required.

@mbolivar-nordic Wasn't the case that the flash_map API should be deprecated?

hakonfam commented on include/flash_pm.h:
```diff
@@ -0,0 +1,46 @@
+		DT_PARENT(DT_PARENT(DT_PARENT(FLASH_AREA_PARTITION(id))))
+#define FLASH_AREA_DEV_LABEL(id)	DT_LABEL(FLASH_AREA_DEV(id))
+
+#endif /* ndef ZEPHYR_INCLUDE_FLASH_PM_H_ */
```
```suggestion
#endif /* ZEPHYR_INCLUDE_FLASH_PM_H_ */
```

de-nordic commented on include/flash_pm.h:
```diff
@@ -0,0 +1,46 @@
+/*
```
Actually I intended to ifdef this file so that when you select alternative PM from Kconfig, FLASH_AREA macros from this file would not be included, instead some user writable header file would be included, similar way the zephyr/include/generated/ files are included during build. This way PM could provide it's own generated file that would contain definitions and FLASH_AREA macros to support them in code. I can do that from flash_map.h

pabigot replied:
>  when you select alternative PM from Kconfig, FLASH_AREA macros from this file would not be included

I don't think that's going to fly, since standard practice now is to define flash areas in devicetree.  If you want a way to *extend* the available flash areas through some sort of partition manager that should be fine, but eliminating the ability to simultaneously specify others in devicetree might be a hard sell.

I suspect it'd be no more than putting:

    #ifdef CONFIG_EXTERNAL_PM
    #include <path/to/generated/pm.h>
    #endif

within the `enum flash_area_partitions {` declaration.

hakonfam replied:
I disagree @pabigot , IMO you pick one or the other. Today we overwrite the partitions defined in device tree when enabling partition manager. This is the only way we get full control of the layout of the device.

pabigot replied:
> This is the only way we get full control of the layout of the device.

I'd need to see much more detail on exactly what the proposed partition manager for upstream Zephyr would look like.  Pending that detail:
* If you want to have a partition manager have full control of the layout of a given device, I can see that making sense.
* Association of a partition manager with one device must not prevent use of immutable device-tree--defined region definitions (as we have today) on another device.

To make an analogy with Linux: It's great to be able to put LVM on a block device so I can split it up for different needs that change over time, but we still need fixed partition tables (MSDOS or GPT), and to be able to put a filesystem on the entire device without a partition table at all.

On initial consideration I would think the flexible way is to continue to define regions in device tree, and specify that a partition manager can further subdivide regions that are assigned to it (which may be the span of an entire device).

This leads to a question of why flash areas aren't simply treated as devices in their own right, delegating to a parent device, but that's something for the future, possibly coordinating with RAM regions.

None of that appears immediately relevant to this PR which AIUI adds a layer of abstraction for retrieving information about a flash region's definition, starting with support for the name/id-based devicetree information we're using today.  But please keep it in mind for whatever follow-on work is envisioned.

hakonfam replied:
> * Association of a partition manager with one device must not prevent use of immutable device-tree--defined region definitions (as we have today) on another device.

Are you talking about multiple devices in the same build (ie cmake invocation) or in separate builds? In a single build it is still possible to "opt out" of partition manager (simply don't enable the USE_PARTITION_MANAGER symbol). If you are talking about another device in another built context, then there is no issue as we decide whether or not to invoke something like partition manager on a per build basis.

> To make an analogy with Linux: It's great to be able to put LVM on a block device so I can split it up for different needs that change over time, but we still need fixed partition tables (MSDOS or GPT), and to be able to put a filesystem on the entire device without a partition table at all.

Note that none of these configurations will change in run time. It can be modified during a firmware update, but I don't consider than runtime. The filesystems (or LVM in this case)
 
> On initial consideration I would think the flexible way is to continue to define regions in device tree, and specify that a partition manager can further subdivide regions that are assigned to it (which may be the span of an entire device).

This is not sufficient as we, as stated above, need full control of the partitions on the device.
DT and PM solve the same part of partitioning, only in different ways.

> This leads to a question of why flash areas aren't simply treated as devices in their own right, delegating to a parent device, but that's something for the future, possibly coordinating with RAM regions.

Are you thinking about a runtime configuration like GPT?

pabigot replied:
I'm saying that if PM is used on the SOC address space, I should still be able to use DT to "partition" an external SPI flash that is a completely different device.

There is nothing in this PR that seems to conflict with that, and I'm fine with this PR once the macros move to `flash_map.h` where the functions that use flash area material generally live.

de-nordic committed:
tests/subsys: Moving tests to new DT_ independent flash API

All references to DT_ macros have been replaced with FLASH_AREA*
API invocations.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
sys/util.h: Add IS_EMPTY and LIST_DROP_EMPTY macros

IS_EMPTY macro allows to check if defined name is empty, i.e.
does not contain replacement list.
LIST_DROP_EMPTY macro may be used to process __VA_ARGS__ type lists,
e.g. a,b,,c , and remove empty elements.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
edtlib: Workaround for generating fixed-partitions compatibles

The script does not allow to generate compatibles that exist in parallel
on two different buses. This caused problem if partitions tables are
placed on flash devices connected via different buses.
The fix enables fixed-partitions compatibles to be present on different
buses at once.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
devicetree: Add DT_FOR_EACH_CHILD macro

The macro iterates through the list of child nodes and invokes provided
macro for each node.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
The work has been abandoned.
